### PR TITLE
FPGA: New Reference design code sample: `fft2d`

### PR DESCRIPTION
**Pull request by yuguen-intel**
# Adding a New Sample(s)

## Description

This change add a new FPGA code sample implementing a 2D FFT on Intel FPGAs.

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): yuguen-intel
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+          half1[i].r() - cos(2 * M_PI * i / kNrPoints) * half2[i].r() -
+          sin(2 * M_PI * i / kNrPoints) * half2[i].i();
+      data[i + kNrPoints / 2].i() =
+          half1[i].i() + sin(2 * M_PI * i / kNrPoints) * half2[i].r() -
+          cos(2 * M_PI * i / kNrPoints) * half2[i].i();
+    }
+
+    free(half1);
+    free(half2);
+  } else {
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:417]: Why the "else { return }?    This is the default behavior from the function, right?

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+     *
+     * In many cases the FFT engine is a building block in a large application.
+     * In this case, the memory layout of the matrix can be altered to achieve
+     * higher memory transfer efficiency. This implementation demonstrates how
+     * an alternative memory layout can improve performance. The host switches
+     * between the two memory layouts using a kernel argument. See the
+     * 'MangleBits' function for additional details.
+     */
+
+    double start_time{};
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:220]: Why the empty brackets?    If you want to specifically initialize the variable to a value you should specify the value.

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    double end_time{};
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:221]:
```suggestion
    double start_time;
    double end_time;
```

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp (hunk 2 above)
Comment (yuguen-intel) [reply DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:220]: True, removing them

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+        magnitude_sum += magnitude;
+        noise_sum += noise;
+      }
+    }
+    double db = 10 * log(magnitude_sum / noise_sum) / log(10.0);
+
+    std::cout << "Signal to noise ratio on output sample: " << db << std::endl;
+    std::cout << " --> " << (db > 120 ? "PASSED" : "FAILED") << std::endl;
+
+    free(input_data, q);
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:330]:
```suggestion
    sycl::free(input_data, q);
```

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+      double tmp = data[i].r();
+      data[i].r() = data[i].i();
+      data[i].i() = tmp;
+    }
+  }
+}
+
+template <int lognr_points>
+void FourierStage(ac_complex<double> *data) {
+  if constexpr (lognr_points > 0) {
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:378]:
```suggestion
  static_assert(lognr_points > 0, "The log of the number of points must be positive");
```

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+      data[i + kNrPoints / 2].i() =
+          half1[i].i() + sin(2 * M_PI * i / kNrPoints) * half2[i].r() -
+          cos(2 * M_PI * i / kNrPoints) * half2[i].i();
+    }
+
+    free(half1);
+    free(half2);
+  } else {
+    return;
+  }
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:419]:
```suggestion
```

File: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp (hunk 1 above)
Comment (yuguen-intel) [reply DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp:417]: Yes, having an assert here makes much more sense

**Review by ishaghosh27 — APPROVED** [review]

**Pull request merged**